libusb/descriptor.c

```diff
@@ -1517,3 +1517,90 @@ void API_EXPORTED libusb_free_interface_association_descriptors(
 		free((void*)iad_array->iad);
 	free(iad_array);
 }
+
+/** \ingroup libusb_desc
+ * Retrieve a device string without needing to open the device.
+ *
+ * \param dev the target device
+ * \param string_type the string type to retrieve
+ * \param data the data buffer for the UTF-8 encoded string.
+ * \param length the size of the data buffer in bytes.  
+ *      USB string descriptors cannot be longer than 
+ *      LIBUSB_DEVICE_STRING_UTF8_BYTES_MAX.
+ * \returns an error code or
+ *      the actual string length in bytes including the null terminator.
+ * \see libusb_get_string_descriptor()
+ * \see libusb_get_string_descriptor_ascii()
+ * 
+ * This function works when the device is still closed since it relies
+ * on the operating system to provide the string.  The operating system
+ * normally reads and caches the common string descriptors during
+ * USB enumeration.
+ *
+ * Since the USB string descriptor could be processed by the OS,
+ * this function returns a UTF-8 encoded string.  The string will
+ * be returned untranslated or in the default OS language
+ * when supported by the OS and USB device.
+ * 
+ * One way to call this function is using a buffer on the stack:
+ * 
+ *     char buffer[LIBUSB_DEVICE_STRING_UTF8_BYTES_MAX];
+ *     int ret = libusb_get_device_string(dev, LIBUSB_DEVICE_STRING_SERIAL_NUMBER, buffer, sizeof(buffer));
+ *     if (ret < 0) {
+ *         // handle error
+ *     }
+ * 
+ * This function is commonly used to get the serial number to allow
+ * for device selection before opening the selected device.
+ */
+int API_EXPORTED libusb_get_device_string(libusb_device* dev,
+	enum libusb_device_string_type string_type, char* data, int length) {
+	char * s;
+	if ((string_type < 0) || (string_type >= LIBUSB_DEVICE_STRING_COUNT)) {
+		return LIBUSB_ERROR_INVALID_PARAM;
+	}
+	if ((NULL != data) && length) {
+		*data = 0;  // return an empty string on errors when possible
+	}
+
+	if (NULL == dev->device_strings_utf8[string_type]) {
+		if (usbi_backend.get_device_string) {
+			s = malloc(LIBUSB_DEVICE_STRING_UTF8_BYTES_MAX);
```

Comment:
since we always preallocate `LIBUSB_DEVICE_STRING_UTF8_BYTES_MAX`, maybe have `dev->device_strings_utf8` as array of fixed-length arrays/strings and omit this dynamic allocation?

Reply:
I originally had it statically allocated, and reconsidered.  I have no idea how annoyed people would be if we suddenly expanded the size of libusb_device by 384 * 3 bytes for a feature no one uses yet.  I figured it was better to be more cautious about object size at the risk of a little malloc performance.  Also, it is very easy to ensure string free when the libusb_device gets freed.

Reply:
I agree, this is minor